src/ucp/proto/proto_perf.c

```diff
@@ -340,6 +334,96 @@ ucs_status_t ucp_proto_perf_add_func(ucp_proto_perf_t *perf, size_t start,
     return status;
 }
 
+ucs_status_t ucp_proto_perf_aggregate(const char *name,
+                                      const ucp_proto_perf_t *const *perf_elems,
+                                      unsigned num_elems,
+                                      ucp_proto_perf_t **perf_p)
+{
+    const ucp_proto_perf_segment_t *seg;
+    ucp_proto_perf_segment_t *new_seg;
+    ucp_proto_perf_t *perf;
+    ucs_list_link_t **pos;
+    ucs_status_t status;
+    size_t start, end;
+    unsigned i;
+    int valid;
+
+    status = ucp_proto_perf_create(name, &perf);
+    if (status != UCS_OK) {
+        goto err;
+    }
+
+    if (num_elems == 0) {
+        goto out;
+    }
+
+    pos = ucs_alloca(sizeof(*pos) * num_elems);
+    for (i = 0; i < num_elems; ++i) {
+        pos[i] = perf_elems[i]->segments.next;
+    }
+
+    start = 0;
+    for (;;) {
```

Comment:
i think we can probably streamline with less nesting like:
```C
i = 0, start = 0, end = INT_MAX;

while (!list_empty(pos[i])) {
  seg = container_of(pos[i]);

  if (seg->end < start) {
    pos[i] = list_next(pos[i]);
    continue;
  }

  if (start < seg->start) {
    start = seg->start;
    end   = seg->end;
    i     = 0;
    continue;
  }

  end = min(seg->end, end);
  if (++i == num_elems) {
    < Create intersection segment >

    start = end + 1;
    end   = INT_MAX;
  }
}

```

Reply:
I would also like to propose the following solution:

```c
    curr            = 0;
    segments[curr]  = ucs_list_head(&perf_elems[curr]->segments,
                                    ucp_proto_perf_segment_t, list);
    while (&segments[curr]->list != &perf_elems[curr]->segments) {
        end = segments[curr]->end;
        for (i = 0; i < num_elems; ++i) {
            if (i == curr) {
                continue;
            }

            segments[i] = ucp_proto_perf_find_segment_lb(perf_elems[i],
                                                         segments[curr]->start);
            if (segments[i] == NULL) {
                goto out;
            }

            if (segments[i]->start > segments[curr]->start) {
                end  = seg->end;
                curr = i;
                i    = 0;
            }

            end = ucs_min(end, seg->end);
        }

        // Create [segments[curr]->start, end] intersection segment

        segments[curr] = ucs_list_next(segments[curr]->list,
                                       ucp_proto_perf_segment_t, list);
    }
```

Reply:
fixed according to @tvegas1 suggestion, since @ivankochin suggestion has more nested loops (ucp_proto_perf_find_segment_lb searches from the start of each perf_elem)

Reply:
I am OK with @tvegas1 solution, but it worth noticing that the problem with `ucp_proto_perf_find_segment_lb` can be solved by passing start pointer to this function.